src/zend/embed.rs

```diff
@@ -0,0 +1,119 @@
+use crate::boxed::ZBox;
+use crate::ffi::{php_embed_init, php_embed_shutdown, zend_eval_stringl, ZEND_RESULT_CODE_SUCCESS};
+use crate::types::{ZendObject, Zval};
+use crate::zend::ExecutorGlobals;
+use parking_lot::{const_rwlock, RwLock, RwLockWriteGuard};
+use std::ffi::{c_char, CString};
+use std::ops::Deref;
+use std::ptr::null_mut;
+
+pub struct Embed;
+
+#[derive(Debug)]
+pub enum EmbedError {
+    InitError,
+    ExecuteError(Option<ZBox<ZendObject>>),
+}
+
+static EMBED_LOCK: RwLock<bool> = const_rwlock(false);
+
+pub struct EmbedWriteGuard<'a, T> {
+    guard: RwLockWriteGuard<'a, bool>,
+    data: T,
+}
+
+impl<'a, T> Deref for EmbedWriteGuard<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.data
+    }
+}
+
+impl Embed {
+    fn init<'a>() -> EmbedWriteGuard<'a, Result<(), EmbedError>> {
+        let mut write_guard = EMBED_LOCK.write();
+
+        if !*write_guard {
+            let result = unsafe { php_embed_init(0, null_mut()) };
+
+            if result != ZEND_RESULT_CODE_SUCCESS {
+                return EmbedWriteGuard {
+                    guard: write_guard,
+                    data: Err(EmbedError::InitError),
+                };
+            }
+
+            *write_guard = true
+        }
+
+        return EmbedWriteGuard {
+            guard: write_guard,
+            data: Ok(()),
+        };
+    }
+
+    // We use a write guard, this should only be called once at a time in non thread safe
+    // In the future we may not need this when it's linked to a thread safe php
+    pub fn run<'a>(code: &str) -> EmbedWriteGuard<'a, Result<Zval, EmbedError>> {
```

Comment:
```suggestion
    pub fn eval<'a>(code: &str) -> EmbedWriteGuard<'a, Result<Zval, EmbedError>> {
```

I would rename to `eval` just to be sure that end user understand what it does